components/config_info/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+            max_value: None,
+            value_options: None,
+            default_value,
+            value_in_file,
+            description,
+        }
+    }
+
+    pub fn set_min_value(mut self, v: ConfigValue<T>) -> Self {
+        self.min_value = Some(v);
+        self
+    }
+
+    pub fn set_max_value(mut self, v: ConfigValue<T>) -> Self {
+        self.max_value = Some(v);
+        self
+    }
+
+    pub fn set_value_options(mut self, ops: Vec<T>) -> Self {
+        self.value_options = Some(ops);
+        self
+    }
+}
+
+#[derive(Serialize)]
+#[serde(untagged)]
+pub enum ConfigValue<T: Serialize> {
+    // a concrete value of type T
+    Concrete(T),
+    // a string description.
+    Desc(String),
+}
+
+#[derive(Serialize)]
+pub enum FieldCfgType {
+    /// represent `[T; N]` and `Vec<T>` types.
+    Array,
+    /// represent `bool` type
+    Boolean,
+    /// represent map types `HashMap<K, V>` or `BTreeMap<K, V>`
+    Map,
+    /// represent all numeric types like `i32`, `u64`, `f64`
+    Number,
+    /// represent all other types that input as string.
+    String,
+}
+
+#[cfg(test)]
+mod tests {
+    use serde::{Deserialize, Serialize};
+    use serde_json::{from_str, to_string, Value};
+    use tikv_util::config::{ReadableDuration, ReadableSize};
+
+    use crate::{self as config_info, ConfigInfo};
+
+    #[derive(Serialize, Deserialize, ConfigInfo)]
+    struct Config {
+        // this is a hidden field that should be skipped in config info.
+        #[config_info(skip)]
+        _hidden: i32,
+        /// This is an i64 field with three integer options,
+        /// the comment can across multiple lines.
+        ///
+        /// But not this line because of the privious empty line.
+        #[config_info(options = r#"[0, 10, 20]"#)]
+        a: i64,
+        #[config_info(min = 0, max = 10)]
+        /// b is an optional i32 field with not-none default value
+        b: Option<i32>,
+        #[config_info(options = r#"["debug", "info"]"#)]
+        /// c is a string value with 2 string options
+        c: String,
+        // a sub module field
+        #[config_info(submodule)]
+        sub: SubConfig,
+        #[config_info(max = "100ms")]
+        /// d is an optional field with default None.
+        d: Option<ReadableDuration>,
+        /// e is a custom type with 2 string options.
+        #[config_info(options = r#"["1MB", "10KB"]"#)]
+        e: ReadableSize,
+        #[config_info(type = "Number", min = 1, max = 100)]
+        /// a custom field with manually assigned type and manually implement of `From`
+        f: NewType,
+        /// This is a field that it's value bound depend on other field or the env variable.
+        #[config_info(
+            default_desc = "MAX(4, CPU * 0.8)",
+            min_desc = "value of `a`",
+            max_desc = "`a` * 5"
+        )]
+        g: u64,
+        /// This is a example of Vec type.
+        h: Vec<u64>,
+        /// This is a example of array type.
```

Comment:
Is it possible to parse other metadata from comment as well?

Reply:
I think so, If we define a special format for each attribute, but there seems no much benefit to do so.

Reply:
I think comment is cleaner and more readable. It would look cumbersome to add yet another modifier other than `online_config`.

Reply:
But document comment is syntax sugar for `#[log = "..."]`, put user-defined attributes to doc comments and manually parse them in the proc-macro looks a bit weird to me.

Reply:
I think parsing doc comment is actual acceptable. Doc tests are generated by parsing code snippet in doc comments. We can introduce our own "dialect" if needed.

In general, I just don't like the idea of adding `#[config_info(..)]` to every config items. Especially when you are putting meaningful information into it.